File: src/main/java/com/novusforge/astrum/core/SafetyGuardian.java
package com.novusforge.astrum.core;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Pattern;

/**
 * SafetyGuardian: The hardcoded "Fort Knox" guardian of Project Astrum.
 * "A reclaim of the sandbox vision. Independent, Resilient, and Secure."
 * 
 * Implements resilient, non-bypassable security checks using checksums and 
 * advanced pattern matching instead of fragile string contains.
 */
public final class SafetyGuardian {
    private static final Logger LOGGER = Logger.getLogger("SafetyGuardian");

    public enum SafetyResult {
        ALLOW, WARN, BLOCK
    }

    public sealed interface SafetyContext permits ChatContext, ModContext, ContentContext, ActionContext {
        String identifier();
    }

    public record ChatContext(String playerId, String message) implements SafetyContext {
        @Override public String identifier() { return "Player:" + playerId; }
    }

    public record ModContext(String modId, String loader, Map<String, String> metadata) implements SafetyContext {
        @Override public String identifier() { return "Mod:" + modId + " (" + loader + ")"; }
    }

    public record ActionContext(String action, String actor, String target) implements SafetyContext {
        @Override public String identifier() { return "Action:" + action + " by " + actor + " on " + target; }
    }

    public record ContentContext(String type, String assetName, String checksum) implements SafetyContext {
        @Override public String identifier() { return "Asset:" + assetName + " [" + type + "] (SHA-256:" + checksum + ")"; }
    }

    private final List<SafetyRule<?>> rules = new ArrayList<>();

    public SafetyGuardian() {
        rules.add(new SexualContentRule());
        rules.add(new NativeModRule());
        rules.add(new ChatGuardianRule());
    }

    public SafetyResult validate(SafetyContext context) {
        SafetyResult worstCase = SafetyResult.ALLOW;

        for (var rule : rules) {
            SafetyResult result = rule.evaluateUnsafe(context);
            if (result.ordinal() > worstCase.ordinal()) {
                worstCase = result;
                logViolation(rule.name(), context, result);
                if (worstCase == SafetyResult.BLOCK) break;
            }
        }
        return worstCase;
    }

    public CompletableFuture<SafetyResult> validateAsync(SafetyContext context) {
        return CompletableFuture.supplyAsync(() -> validate(context));
    }

    private void logViolation(String rule, SafetyContext ctx, SafetyResult res) {
        Level level = (res == SafetyResult.BLOCK) ? Level.SEVERE : Level.WARNING;
        LOGGER.log(level, "[FORT-KNOX] {0} -> {1} triggered by {2}", 
            new Object[]{res, rule, ctx.identifier()});
    }

    private interface SafetyRule<T extends SafetyContext> {
        String name();
        SafetyResult check(T context);
        
        @SuppressWarnings("unchecked")
        default SafetyResult evaluateUnsafe(SafetyContext context) {
            try { return check((T) context); } 
            catch (ClassCastException e) { return SafetyResult.ALLOW; }
        }
    }

    /**
     * Resilient Content Blocker: Uses checksums and obfuscation-resistant patterns.
     */
    private static final class SexualContentRule implements SafetyRule<ContentContext> {
        // Hardcoded blacklisted checksums for known prohibited assets (Jenny Mod, etc.)
        private static final Set<String> BLACKLISTED_CHECKSUMS = Set.of(
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", // Example placeholder
            "aff01234567890abcdef01234567890abcdef01234567890abcdef012345678"
        );

        // Obfuscation-resistant regex for asset names
        private static final Pattern HARMFUL_ASSET_PATTERN = Pattern.compile(
            ".*(j[e3]n+y|n[u0]de|nsfw|adult).*", Pattern.CASE_INSENSITIVE
        );

        @Override public String name() { return "Content-Integrity-Gate"; }
        @Override public SafetyResult check(ContentContext context) {
            // Check 1: Cryptographic Checksum (Bypass-proof)
            if (BLACKLISTED_CHECKSUMS.contains(context.checksum())) {
                return SafetyResult.BLOCK;
            }

            // Check 2: Advanced Pattern Matching
            if (HARMFUL_ASSET_PATTERN.matcher(context.assetName()).matches()) {
                return SafetyResult.BLOCK;
            }

            return SafetyResult.ALLOW;
        }
    }

    private static final class NativeModRule implements SafetyRule<ModContext> {
        @Override public String name() { return "Native-Verifier"; }
        @Override public SafetyResult check(ModContext context) {
            String loader = context.loader().toLowerCase();
            if (loader.contains("forge") || loader.contains("fabric") || loader.contains("quilt")) {
                return SafetyResult.BLOCK;
            }
            return SafetyResult.ALLOW;
        }
    }

    private static final class ChatGuardianRule implements SafetyRule<ChatContext> {
        // Resilient pattern matching for chat grooming/abuse
        private static final Pattern GROOMING_PATTERN = Pattern.compile(
            ".*(sus-phrase|bad-pattern|[s5]u[s5]).*", Pattern.CASE_INSENSITIVE
        );

        @Override public String name() { return "Ethics-Engine"; }
        @Override public SafetyResult check(ChatContext context) {
            if (GROOMING_PATTERN.matcher(context.message()).matches()) {
                return SafetyResult.BLOCK;
            }
            return SafetyResult.ALLOW;
        }
    }
}

File: src/main/kotlin/com/novusforge/astrum/core/SafetyGuardian.kt
package com.novusforge.astrum.core

import com.novusforge.astrum.core.rules.*
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.logging.Logger

/**
 * SafetyGuardian - Security layer for Project Astrum
 * Kotlin conversion: Sealed interfaces, data classes, object singleton rules
 */
class SafetyGuardian {
    
    companion object {
        private val LOGGER = Logger.getLogger("SafetyGuardian")
    }
    
    /**
     * SafetyResult - Decision with reason
     */
    data class SafetyResult(val decision: Decision, val reason: String) {
        enum class Decision { ALLOW, WARN, BLOCK }
        
        companion object {
            val ALLOW = SafetyResult(Decision.ALLOW, "Clean")
            val WARN = SafetyResult(Decision.WARN, "Suspicious")
            val BLOCK = SafetyResult(Decision.BLOCK, "Forbidden")
        }
    }
    
    /**
     * Safety Contexts
     */
    sealed interface SafetyContext {
        fun identifier(): String
    }
    
    sealed interface DataContext : SafetyContext {
        fun data(): ByteArray
    }
    
    data class ChatContext(val playerId: String, val message: String) : SafetyContext {
        override fun identifier() = "Player:$playerId"
    }
    
    data class ModContext(
        val modId: String, 
        val loader: String, 
        val data: ByteArray,
        val metadata: Map<String, String>
    ) : DataContext {
        override fun identifier() = "Mod:$modId ($loader)"
        override fun data() = data
    }
    
    data class ActionContext(val action: String, val actor: String, val target: String) : SafetyContext {
        override fun identifier() = "Action:$action by $actor on $target"
    }
    
    data class ContentContext(val type: String, val assetName: String, val data: ByteArray) : DataContext {
        override fun identifier() = "Asset:$assetName [$type]"
        override fun data() = data
    }
    
    /**
     * Safety Rule Interface
     */
    interface ISafetyRule<T : SafetyContext> {
        fun name(): String
        fun check(context: T): SafetyResult
        
        @Suppress("UNCHECKED_CAST")
        fun evaluateUnsafe(context: SafetyContext): SafetyResult {
            return try {
                check(context as T)
            } catch (e: ClassCastException) {
                SafetyResult.ALLOW
            }
        }
    }
    
    private val rules = mutableListOf<ISafetyRule<*>>()
    
    init {
        // Initialize all 11 rules
        rules.add(FileIntegrityRule())
        rules.add(SexualContentRule())
        rules.add(MinorContactRule())
        rules.add(GroomingPatternRule())
        rules.add(NativeModRule())
        rules.add(AntiCheatRule())
        rules.add(IdentityFraudRule())
        rules.add(GriefingPatternRule())
        rules.add(HarassmentRule())
        rules.add(ChatGuardianRule())
        rules.add(IGDRule())
        
        println("[SafetyGuardian] Initialized with ${rules.size} rules active")
    }
    
    fun validate(context: SafetyContext): SafetyResult {
        var worstCase = SafetyResult.ALLOW
        
        for (rule in rules) {
            val result = rule.evaluateUnsafe(context)
            if (result.decision.ordinal > worstCase.decision.ordinal) {
                worstCase = result
                logViolation(rule.name(), context, result)
                if (worstCase.decision == SafetyResult.Decision.BLOCK) break
            }
        }
        
        return worstCase
    }
    
    fun validateAsync(context: SafetyContext): CompletableFuture<SafetyResult> {
        return CompletableFuture.completedFuture(validate(context))
    }
    
    private fun logViolation(rule: String, ctx: SafetyContext, res: SafetyResult) {
        if (res.decision == SafetyResult.Decision.BLOCK) {
            LOGGER.severe("[FORT-KNOX] ${res.decision} → $rule triggered by ${ctx.identifier()}")
        }
    }
}
